src/V3EmitCSyms.cpp

```diff
@@ -661,293 +615,169 @@ void EmitCSyms::emitSymImpPreamble() {
+        stmt += " = &";
+        stmt += VIdProtect::protectIf(scopep->nameDotless(), scopep->protect()) + ";";
+        add(stmt);
     }
 
-    puts("// Setup each module's pointer back to symbol table (for public functions)\n");
+    add("// Setup each module's pointer back to symbol table (for public functions)");
     for (const ScopeModPair& i : m_scopes) {
         const AstScope* const scopep = i.first;
         AstNodeModule* const modp = i.second;
-        checkSplit(false);
         // first is used by AstCoverDecl's call to __vlCoverInsert
         const bool first = !modp->user1();
         modp->user1(true);
-        putns(scopep, VIdProtect::protectIf(scopep->nameDotless(), scopep->protect()) + "."
-                          + protect("__Vconfigure") + "(" + (first ? "true" : "false") + ");\n");
-        ++m_numStmts;
+        add(VIdProtect::protectIf(scopep->nameDotless(), scopep->protect()) + "."
+            + protect("__Vconfigure") + "(" + (first ? "true" : "false") + ");");
     }
 
-    if (!m_scopeNames.empty()) {  // Setup scope names
-        puts("// Setup scopes\n");
-        for (const auto& itpair : m_scopeNames) {
-            checkSplit(false);
-            const ScopeData& sd = itpair.second;
-            putns(sd.m_nodep, protect("__Vscope_" + sd.m_symName) + ".configure(this, name(), ");
-            putsQuoted(VIdProtect::protectWordsIf(sd.m_prettyName, true));
-            puts(", ");
-            putsQuoted(protect(scopeDecodeIdentifier(sd.m_prettyName)));
-            puts(", ");
-            putsQuoted(sd.m_defName);
-            puts(", ");
-            puts(std::to_string(sd.m_timeunit));
-            puts(", VerilatedScope::" + sd.m_type + ");\n");
-            ++m_numStmts;
-        }
+    add("// Setup scopes");
+    for (const auto& itpair : m_scopeNames) {
+        const ScopeData& sd = itpair.second;
+        std::string stmt;
+        stmt += protect("__Vscope_" + sd.m_symName) + ".configure(this, name(), \"";
+        stmt += V3OutFormatter::quoteNameControls(
+            VIdProtect::protectWordsIf(sd.m_prettyName, true));
+        stmt += "\", \"";
+        stmt += V3OutFormatter::quoteNameControls(protect(scopeDecodeIdentifier(sd.m_prettyName)));
+        stmt += "\", \"";
+        stmt += V3OutFormatter::quoteNameControls(sd.m_defName);
+        stmt += "\", ";
+        stmt += std::to_string(sd.m_timeunit);
+        stmt += ", VerilatedScope::" + sd.m_type + ");";
+        add(stmt);
     }
 
-    emitScopeHier(false);
+    emitScopeHier(stmts, false);
 
-    // Everything past here is in the __Vfinal loop, so start a new split file if needed
-    closeSplit();
+    if (!v3Global.dpi()) return stmts;
 
-    if (v3Global.dpi()) {
-
-        puts("// Setup export functions\n");
-        puts("for (int __Vfinal = 0; __Vfinal < 2; ++__Vfinal) {\n");
+    for (const std::string vfinal : {"0", "1"}) {
+        add("// Setup export functions - final: " + vfinal);
```

Comment:
I appreciate this loop used to be at run-time, so we are generating twice as much code for these. I think this is still preferable as it can be split up easily now.

Reply:
Note also could just rewrite the run-time to not need to do this twice, varInset already ignores the !final pass.